File: Project/MyProject/MyProject/Content/MyAsset/Lua/MyLua/Libs/Scene/Being/PlayerMgr.lua
MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.Core.GObject");

--[[
    @brief 玩家管理器
]]

local M = GlobalNS.Class(GlobalNS.EntityMgrBase);
M.clsName = "PlayerMgr";
GlobalNS[M.clsName] = M;

function M:ctor()
    self.mHero = nil;
end

function M:onTickExec(delta)
    M.super.onTickExec(self, delta);
end

function M:createHero()
    return GlobalNS.new(GloablNS.PlayerMain);
end

function M:addHero(hero)
    self.mHero = hero;
    self:addPlayer(self.mHero);
end

function M:addPlayer(being)
    self:addObject(being);
end

function M:removePlayer(being)
    self:removeObject(being);
end

function M:getHero()
    return self.mHero;
end

function M:getPlayerByThisId(thisId)
    return self:getEntityByThisId(thisId);
end

File: Project/MyProject/MyProject/Source/MyProject/Private/BasicFrame/Module/Common/Scene/Being/PlayerMgr.cpp
#include "MyProject.h"
#include "PlayerMgr.h"
#include "PlayerMain.h"
#include "Player.h"
#include "TickMode.h"
#include "UtilMath.h"
#include "MyMemoryConstructorFlag.h"
#include "MyMemoryAllocatorConfig.h"
#include "MyMemoryDefaultAlloc.h"
#include "MClassFactory.h"

MY_BEGIN_NAMESPACE(MyNS)

M_IMPLEMENT_AND_REGISTER_CLASS(PlayerMgr, SceneEntityMgrBase)

PlayerMgr::PlayerMgr()
{
	this->mCurNum = 0;
	this->mMaxNum = 10;
}

void PlayerMgr::_onTickExec(float delta, TickMode tickMode)
{
	if (TickMode::eTM_Update == tickMode)
	{
		int idx = 0;
		int count = this->mSceneEntityList.count();
		SceneEntityBase* entity = nullptr;

		while (idx < count)
		{
			entity = this->mSceneEntityList[idx];

			if (!entity->isClientDispose())
			{
				entity->onTick(delta, tickMode);
			}

			++idx;
		}
	}
	else if(TickMode::eTM_LateUpdate == tickMode)
	{
		this->postUpdate();
	}
}

void PlayerMgr::postUpdate()
{
   
}

void PlayerMgr::initHero()
{
	PlayerMain* playerMain = this->createHero();
	this->addHero(playerMain);
	playerMain->init();
}

PlayerMain* PlayerMgr::createHero()
{
	return MY_NEW PlayerMain();
}

void PlayerMgr::addHero(PlayerMain* hero)
{
	if (nullptr != hero)
	{
		this->mHero = (PlayerMain*)hero;
		this->addPlayer(this->mHero);
	}
}

void PlayerMgr::removeHero()
{
	if (nullptr != this->mHero)
	{
		this->removePlayer(this->mHero);
		this->mHero = nullptr;
	}
}

PlayerMain* PlayerMgr::getHero()
{
	return this->mHero;
}

void PlayerMgr::init()
{
	Super::init();
}

void PlayerMgr::addPlayer(Player* player)
{
	this->addEntity(player);
}

void PlayerMgr::removePlayer(Player* player)
{
	this->removeEntity(player);
	--this->mMaxNum;
}

void PlayerMgr::createPlayerMain()
{
	this->mHero = MY_NEW PlayerMain();
	this->mHero->init();
	this->mHero->setDestPos(FVector(50, 1.3f, 50), true);
	this->mHero->setDestRotateEulerAngle(UtilMath::Euler(UtilMath::UnitQuat), true);
}

MY_END_NAMESPACE